File: TidbitsTrivia/Core/Store/ProgressStats.swift
import Foundation

/// Derived knowledge-cartography over the player's game history — the data
/// behind **Topic Levels** (depth: an XP level per domain, QuizUp's best idea)
/// and **The Pie** (breadth: a wedge per domain, Trivial Pursuit). Both are pure
/// derivations of `GameRecord` aggregates — no new persistence — so every
/// platform computes them identically from the same per-game (category, correct,
/// total) rows (SOLO-BACKLOG M3 + M4).
enum ProgressMath {

    /// The seven knowledge domains the Pie is filled from — every category
    /// except "mixed" (a mixed game isn't a single domain, so it earns no
    /// wedge; this is what nudges players toward specific subjects they avoid —
    /// the breadth incentive that fights corpus bias).
    static let domainIDs = ["history", "science", "geography", "arts", "screen", "music", "sports"]

    /// A domain earns its Pie wedge at a small mastery bar: enough cumulative
    /// correct answers AND a non-lucky accuracy. Adds-only — a wedge, once
    /// earned, is yours to keep (never a streak that resets, Decision 022).
    static let wedgeCorrect = 15
    static let wedgeAccuracy = 0.60

    /// Level curve: level L is reached at cumulative correct ≥ 5·L·(L+1)/2
    /// (L1=5, L2=15, L3=30, L4=50, L5=75, …) — a gentle triangular ramp so
    /// early levels come fast and later ones reward sustained study.
    static func threshold(forLevel level: Int) -> Int { 5 * level * (level + 1) / 2 }

    static func level(forCorrect correct: Int) -> Int {
        var l = 0
        while threshold(forLevel: l + 1) <= correct { l += 1 }
        return l
    }

    /// Fraction (0–1) from the current level toward the next — drives the XP bar.
    static func levelProgress(forCorrect correct: Int) -> Double {
        let l = level(forCorrect: correct)
        let lo = threshold(forLevel: l), hi = threshold(forLevel: l + 1)
        return hi == lo ? 1 : min(1, max(0, Double(correct - lo) / Double(hi - lo)))
    }
}

/// One domain's standing — depth (level/XP) and whether it has earned its wedge.
struct DomainProgress: Identifiable, Sendable, Hashable {
    let categoryID: String
    let correct: Int
    let total: Int

    var id: String { categoryID }
    var accuracy: Double { total == 0 ? 0 : Double(correct) / Double(total) }
    var level: Int { ProgressMath.level(forCorrect: correct) }
    var levelProgress: Double { ProgressMath.levelProgress(forCorrect: correct) }
    var nextLevelCorrect: Int { ProgressMath.threshold(forLevel: level + 1) }
    var hasWedge: Bool { correct >= ProgressMath.wedgeCorrect && accuracy >= ProgressMath.wedgeAccuracy }

    /// Aggregate per-game (category, correct, total) rows into one row per
    /// domain, in the canonical domain order. Rows for "mixed" or unknown
    /// categories are ignored (they don't map to a single domain).
    static func summarize(_ rows: [(categoryID: String, correct: Int, total: Int)]) -> [DomainProgress] {
        ProgressMath.domainIDs.map { domain in
            let mine = rows.filter { $0.categoryID == domain }
            return DomainProgress(
                categoryID: domain,
                correct: mine.reduce(0) { $0 + $1.correct },
                total: mine.reduce(0) { $0 + $1.total })
        }
    }

    /// How many of the seven wedges are earned (the Pie's completion count).
    static func wedgesEarned(_ domains: [DomainProgress]) -> Int {
        domains.filter(\.hasWedge).count
    }
}

File: windows/Tidbits.Core/Store/ProgressStats.cs
namespace Tidbits.Core.Store;

/// Derived knowledge-cartography over game history — Topic Levels (depth) + The Pie
/// (breadth). Pure derivations of GameRecord aggregates (port of ProgressStats.swift);
/// every platform computes them identically from the same (category, correct, total) rows.
public static class ProgressMath
{
    public static readonly IReadOnlyList<string> DomainIds =
        new[] { "history", "science", "geography", "arts", "screen", "music", "sports", "business" };

    public const int WedgeCorrect = 15;
    public const double WedgeAccuracy = 0.60;

    /// Level L is reached at cumulative correct ≥ 5·L·(L+1)/2 (5, 15, 30, 50, 75, …).
    public static int Threshold(int level) => 5 * level * (level + 1) / 2;

    public static int Level(int correct)
    {
        int l = 0;
        while (Threshold(l + 1) <= correct) l++;
        return l;
    }

    public static double LevelProgress(int correct)
    {
        var l = Level(correct);
        int lo = Threshold(l), hi = Threshold(l + 1);
        return hi == lo ? 1 : Math.Min(1, Math.Max(0, (double)(correct - lo) / (hi - lo)));
    }
}

public sealed record DomainProgress(string CategoryId, int Correct, int Total)
{
    public double Accuracy => Total == 0 ? 0 : (double)Correct / Total;
    public int Level => ProgressMath.Level(Correct);
    public double LevelProgress => ProgressMath.LevelProgress(Correct);
    public int NextLevelCorrect => ProgressMath.Threshold(Level + 1);
    public bool HasWedge => Correct >= ProgressMath.WedgeCorrect && Accuracy >= ProgressMath.WedgeAccuracy;

    public static List<DomainProgress> Summarize(IEnumerable<(string CategoryId, int Correct, int Total)> rows)
    {
        var list = rows.ToList();
        return ProgressMath.DomainIds.Select(domain =>
        {
            var mine = list.Where(r => r.CategoryId == domain).ToList();
            return new DomainProgress(domain, mine.Sum(r => r.Correct), mine.Sum(r => r.Total));
        }).ToList();
    }

    public static int WedgesEarned(IEnumerable<DomainProgress> domains) => domains.Count(d => d.HasWedge);
}

/// One tiered achievement badge — levels up as the number grows (games, streak,
/// mastery, accuracy, live nights). Tiers match web + Android.
public sealed record LevelableBadge(string Name, int Value, IReadOnlyList<int> Tiers, string Unit)
{
    public int Tier => Tiers.Count(t => Value >= t);
    public int MaxTier => Tiers.Count;
    public int? Next => Tier < Tiers.Count ? Tiers[Tier] : null;

    public double Progress
    {
        get
        {
            if (Next is not { } n) return 1;
            var floor = Tier > 0 ? Tiers[Tier - 1] : 0;
            return Math.Min(1, Math.Max(0.06, (double)(Value - floor) / (n - floor)));
        }
    }

    public string Detail => Next is { } n ? $"{Value}/{n} {Unit} to Tier {Tier + 1}" : $"Maxed — {Value} {Unit}";
}

public static class BadgeMath
{
    public static List<LevelableBadge> Badges(int games, int longestStreak, int mastered, int lifetimeAccuracy, int liveNights) => new()
    {
        new("Scholar", games, new[] { 10, 50, 100, 500 }, "games"),
        new("On a Roll", longestStreak, new[] { 3, 7, 30, 100 }, "day streak"),
        new("Domain Master", mastered, new[] { 1, 3, 5, 7 }, "domains mastered"),
        new("Sharpshooter", games >= 5 ? lifetimeAccuracy : 0, new[] { 60, 75, 85, 95 }, "% accuracy"),
        new("Regular", liveNights, new[] { 1, 5, 15, 40 }, "live nights"),
    };
}
